reasoner/planner/AnswerCountEstimator.java

```diff
@@ -0,0 +1,486 @@
+                Set<Variable> filteredVariablesInEstimate = estimate.variables.stream()
+                        .filter(variableFilter::contains).collect(Collectors.toSet());
+
+                long currentCostToCover = costOfEstimateCover(filteredVariablesInEstimate, costCover);
+                if (currentCostToCover > estimate.answerEstimate(filteredVariablesInEstimate)) {
+                    filteredVariablesInEstimate.forEach(v -> costCover.put(v, estimate));
+                }
+            }
+            return costCover;
+        }
+
+        private static long costOfEstimateCover(Set<Variable> variablesToConsider, Map<Variable, LocalEstimate> coverMap) {
+            Set<LocalEstimate> subsetCoveredBy = coverMap.keySet().stream().filter(variablesToConsider::contains)
+                    .map(coverMap::get).collect(Collectors.toSet());
+            return subsetCoveredBy.stream().map(estimate -> estimate.answerEstimate(variablesToConsider)).reduce(1L, (x, y) -> x * y);
+        }
+
+        public boolean mayInitialize() {
+            return initializationStatus == InitializationStatus.NOT_STARTED || initializationStatus == InitializationStatus.RESET;
+        }
+
+        private void reset() {
+            assert this.initializationStatus == InitializationStatus.COMPLETE;
+            this.initializationStatus = InitializationStatus.RESET;
+            this.inferrableEstimates = null;
+            this.fullAnswerCount = -1;
+            this.negatedsCost = -1;
+        }
+
+        private void registerTriggeredRules() {
+            resolvables().filter(this::isInferrable).map(Resolvable::asConcludable).forEachRemaining(concludable -> {
+                Map<Rule, Set<Unifier>> unifiers = logicMgr.applicableRules(concludable);
+                iterate(unifiers.keySet()).forEachRemaining(rule -> {
+                    answerCountEstimator.registerConjunctionAndInitialize(rule.condition().conjunction());
+                });
+            });
+        }
+
+        private void initialize() {
+            if (!this.mayInitialize()) return;
+            //TODO: How does traversal handle type Variables?
+
+            initializationStatus = InitializationStatus.IN_PROGRESS;
+            // Create a baseline estimate from type information
+            if (unaryEstimateCover == null) unaryEstimateCover = computeNonInferredUnaryEstimateCover();
+
+            if (retrievableEstimates == null) retrievableEstimates = deriveEstimatesFromRetrievables();
+
+            registerTriggeredRules();
+
+            if (inferrableEstimates == null) {
+                inferrableEstimates = deriveEstimatesFromInferrables();
+                unaryEstimates = computeUnaryEstimatesWithInference(unaryEstimateCover);
+                unaryEstimateCover = computeFinalUnaryEstimateCover(unaryEstimateCover); // recompute with inferred
+            }
+
+            if (this.negatedsCost == -1) this.negatedsCost = computeNegatedsCost();
```

Comment:
interestingly, by the time we get there, these should always be null? because we only get past the initial `if` when the state is the first state or the RESET state.

We acutally mostly want to make sure that we only do the `unaryEstimateCover` and `retrievableEstimates` once, because they aren't affected by recursion! So we can protect that with the `state == NOT_STARTED/EMPTY`!

Reply:
That would move the setting of InitializationState = IN_PROGRESS to somewhere in the middle of the function which is a bit weird for tracking what's going on. I could use a temporary flag, but the ==null check was used in a lot of places in reasoner to enforce that it's computed only once and then cached .